_posts/2019-08-12-https-nginx.md

```diff
@@ -0,0 +1,209 @@
+
+In order to run this command, we need to get certbot installed in the nginx container, so we add this line to the Dockerfile
+```
+RUN apt-get install -y certbot python-certbot-nginx 
+```
+
+Then, we rebuild the image and run a container from it. At this point we installed the certbot agent. We now run the command in the container spawned from this image:
+```bash
+  $ docker ps
+  $ docker exec -it <NGINX_CONTAINER_ID> bash
+  $ letsencrypt certonly -a webroot --webroot-path=/var/www/folder_where_challenge_will_be_put -d www.domain.co.uk
+```
+
+If we run this command, the challenge will fail. As we said previously, the challenge consists in retrieving a file that is created on our machine. We need to change the nginx config to serve the folder where the file has been created, in our example it's `/var/www/folder_where_challenge_will_be_put/`. 
+We need to add this part to the config file:
+
+```
+  location ~ /.well-known {
+    allow all;
+    root /var/www/folder_where_challenge_will_be_put/;
+  }
+```
+
+If the last command succeeds, we generated the certificates which have been installed at `/etc/letsencrypt/live/www.domain.co.uk/`.
+
+You can find the code of this step [here](https://github.com/nabilbouab/tech-blog-labs/tree/step-2)
+
+
+## Step 3: Make HTTPS work
+
+This last step will consist in two sub steps:
+  * We first need to make a redirection from HTTP to HTTPS
+  * We need to make NGINX do the encryption/decryption by providing it access to the generated certificates
+
+![SSL termination diagram](/assets/nginx-ssl-termination.png)
+
+In order to do the ssl termination, NGINX needs to be able to access the certificates which will be used to encrypt and decrypt the data. We need to change the NGINX config to point it to the newly generated certificates and make the redirection that way:
+
+```
+worker_processes  1;
+
+events {
+	worker_connections  1024;
+}
+
+
+http {
+	include       mime.types;
+	default_type  application/octet-stream;
+
+	sendfile        on;
+
+	keepalive_timeout  65;
+
+
+	upstream public {
+		server web:3000;
+	}
+
+	server {
+		listen 80;
+		server_name our-awesome-domain.fr;
+
+		location ~ /.well-known {
+			allow all;
+			root /var/www/our-awesome-domain.fr/;
+		}
+
+    location /favicon.ico {
+      root /var/www/;
+    }
+
+    location / {
+      return 301 https://$host$request_uri;
+    }
+	}
+
+  server {
+    listen 443 ssl;
+    server_name our-awesome-domain.fr;
+
+    ssl_certificate /etc/letsencrypt/live/www.our-awesome-domain.fr/fullchain.pem;
+    ssl_certificate_key /etc/letsencrypt/live/www.our-awesome-domain.fr/privkey.pem;
+
+    ssl_session_cache shared:SSL:50m;
+    ssl_session_timeout 10m;
+
+    # ssl_dhparam /etc/ssl/dhparams/dhparam.pem;
```

Comment:
This is quite dangerous, by default NGINX uses a 1024 bit prime for the Diffie-Hellman key exchange which is more or less broken (and our friends at the NSA have actually been using it quite a lot, see some discussion here: [
How is NSA breaking so much crypto?](https://freedom-to-tinker.com/2015/10/14/how-is-nsa-breaking-so-much-crypto/) ). It is recommended to generate your own DH params of at least 2048 bits.

While I understand that this is not necessarily the subject of this article, I would uncomment this line and explain briefly what it does (because I know that some people will copy/paste that in their own nginx.conf and we would spread an vulnerability!)